### PR TITLE
Partially revert "CreateCgroupPath: only enable needed controllers"

### LINKED ISSUE #2394: `runc update` broken in cgroup v2
If we update a resource which did not limited in the beginning, it will
have no effective.
For example:
We use default `config.json` to create a container, and then run:
```
runc update --cpu-quota 1 test
runc update --cpuset-cpus 1 test
```
It has no error input, but actually it has no effective.

### PR DESCRIPTION
**Pull request by lifubang**
fix #2394 

1. Partially revert "CreateCgroupPath: only enable needed controllers"
Because if we update a resource which did not limited in the beginning, it will
have no effective.
2. Returns err if we use an non enabled controller, or else the user may feel success, but actually
there are no effective.

For #2367 , I think we don't need fully revert https://github.com/opencontainers/runc/commit/4b4bc995ad4ee4679ffbe38f1672be01c24256fc , because rootless need some part in it to check errors.

Signed-off-by: lifubang <lifubang@acmcoder.com>

### CODE REVIEW

File: libcontainer/cgroups/fs2/fs2.go [hunk 1]
```diff
@@ -189,57 +189,41 @@ func (m *manager) Set(container *configs.Config) error {
 	if err := m.getControllers(); err != nil {
 		return err
 	}
-	var errs []error
```
Comment (AkihiroSuda) [inline libcontainer/cgroups/fs2/fs2.go:192]: why?

Comment (lifubang) [reply libcontainer/cgroups/fs2/fs2.go:192]: I think if there is an error, we should return without try other controllers.

Do you mean we should let enabled controllers success when we update more than 1 controller in one command, for example:
```
runc update --cpuset-cpus 1 --memory 5m test
```
We should let memory limit set success?

Comment (lifubang) [reply libcontainer/cgroups/fs2/fs2.go:192]: In v1, it doesn't retry any other subsystems when got an error in one subsystem.

**Comment by lifubang** [comment]
CI error:
```
# time="2020-05-10T02:06:34Z" level=error msg="container_linux.go:348: starting container process caused: process_linux.go:438: container init caused: process_linux.go:404: setting cgroup config for procHooks process caused: failed to write \"33554432\" to \"/sys/fs/cgroup/runc-cgroups-integration-test/test-cgroup/memory.max\": open /sys/fs/cgroup/runc-cgroups-integration-test/test-cgroup/memory.max: permission denied"
```
I think this error is ignored in the past.
So, I think we need #2390 to fix this error.

File: libcontainer/cgroups/fs2/memory.go [hunk 2]
```diff
@@ -50,8 +50,10 @@ func setMemory(dirPath string, cgroup *configs.Cgroup) error {
 		// memory and memorySwap set to the same value -- disable swap
 		swapStr = "0"
 	}
-	if err := fscommon.WriteFile(dirPath, "memory.swap.max", swapStr); err != nil {
-		return err
+	if swapStr != "" {
```
Comment (lifubang) [inline libcontainer/cgroups/fs2/memory.go:53]: CI error:
```
# time="2020-05-10T01:05:29Z" level=error msg="container_linux.go:348: starting container process caused: process_linux.go:438: container init caused: process_linux.go:404: setting cgroup config for procHooks process caused: failed to write \"\" to \"/sys/fs/cgroup/runc-cgroups-integration-test/test-cgroup/memory.swap.max\": open /sys/fs/cgroup/runc-cgroups-integration-test/test-cgroup/memory.swap.max: permission denied"
```
This error is introduced by #2370 . So we need to fix it in here.

**Comment by lifubang** [comment]
Waiting for #2390 complete.

File: libcontainer/cgroups/fs2/create.go [hunk 3]
```diff
@@ -10,7 +11,7 @@ import (
 	"github.com/opencontainers/runc/libcontainer/configs"
 )
 
-// neededControllers returns the string to write to cgroup.subtree_control,
+// neededControllers returns the minimal string that should be write to cgroup.subtree_control,
```
Comment (kolyshkin) [inline libcontainer/cgroups/fs2/create.go:14]: `string to write to` was OK

if you want `should be`, it should be `written`.

File: libcontainer/cgroups/fs2/memory.go (hunk 2 above)
Comment (kolyshkin) [reply libcontainer/cgroups/fs2/memory.go:53]: This is a separate fix and it requires a separate patch.

Also, why haven't we caught this before?

File: libcontainer/cgroups/fs2/create.go (hunk 3 above)
Comment (kolyshkin) [reply libcontainer/cgroups/fs2/create.go:14]: if you're not using neededControllers, why keep it?

Comment (lifubang) [reply libcontainer/cgroups/fs2/create.go:14]: Rootless needs it.
https://github.com/opencontainers/runc/blob/744bbe8128754ecb30a18465281ec3e02ee0766f/libcontainer/cgroups/fs2/fs2.go#L77-L85
How about move it to fs2.go and change the implementation way.

Comment (kolyshkin) [reply libcontainer/cgroups/fs2/create.go:14]: Yes please.

**Comment by kolyshkin** [comment]
@lifubang #2390 is now merged

**Comment by lifubang** [comment]
How about modify this check:
https://github.com/opencontainers/runc/blob/4e2499e8de399f0cf281825d2e651d14ba58b21c/tests/integration/update.bats#L84-L87
Chang to check whether men controller is enabled or not in /sys/fs/cgroup/cgroups.controller?

File: libcontainer/cgroups/fs2/fs2.go (hunk 1 above)
Comment (kolyshkin) [reply libcontainer/cgroups/fs2/fs2.go:178]: make sense.

File: libcontainer/cgroups/fs2/create.go [hunk 4]
```diff
@@ -10,24 +11,27 @@ import (
...
-// containing the list of controllers to enable (for example, "+cpu +pids"),
+func supportedControllers(cgroup *configs.Cgroup) ([]byte, error) {
+	const file = UnifiedMountpoint + "/cgroup.controllers"
+	return ioutil.ReadFile(file)
+}
+
+// neededControllers returns the number of controllers to enable,
 // based on (1) controllers available and (2) resources that are being set.
 //
 // The resulting string does not include "pseudo" controllers such as
```
Comment (kolyshkin) [inline libcontainer/cgroups/fs2/create.go:22]: we don't have a string anymore so the comment needs to be fixed

File: libcontainer/cgroups/fs2/create.go [hunk 5]
```diff
@@ -10,24 +11,27 @@ import (
 	"github.com/opencontainers/runc/libcontainer/configs"
 )
 
-// neededControllers returns the string to write to cgroup.subtree_control,
-// containing the list of controllers to enable (for example, "+cpu +pids"),
+func supportedControllers(cgroup *configs.Cgroup) ([]byte, error) {
```
Comment (kolyshkin) [inline libcontainer/cgroups/fs2/create.go:14]: I suggest to return `bool` here as this is what the caller needs.

File: libcontainer/cgroups/fs2/create.go [hunk 6]
```diff
@@ -10,24 +11,27 @@ import (
...
+}
+
+// neededControllers returns the number of controllers to enable,
 // based on (1) controllers available and (2) resources that are being set.
 //
 // The resulting string does not include "pseudo" controllers such as
 // "freezer" and "devices".
-func neededControllers(cgroup *configs.Cgroup) ([]string, error) {
-	var list []string
+func neededControllers(cgroup *configs.Cgroup) (int, error) {
```
Comment (kolyshkin) [inline libcontainer/cgroups/fs2/create.go:24]: Maybe return `bool` here as no one is interested in the number of controllers to be enabled, but rather whether any controllers need to be enabled or not.

Comment (kolyshkin) [reply libcontainer/cgroups/fs2/create.go:24]: You can also optimize things by returning true as soon as there is a controller that needs to be enabled.

Something like

```go
￼	// 
￼	have := func(controller string) bool {
￼		if _, ok := avail[controller]; ok {
￼	￼	￼	return true
￼		}
￼		return false
￼	}

...

if isPidsSet(cgroup) && have("pids") {
￼		return true
}
if isMemorySet(cgroup) && have("memory") {
    return true
}
....
```
		add("memory")
	}
		add("pids")

Comment (kolyshkin) [reply libcontainer/cgroups/fs2/create.go:24]: And maybe rename to `needAnyControllers()` or smth similar

File: libcontainer/cgroups/fs2/fs2.go [hunk 7]
```diff
@@ -175,57 +175,41 @@ func (m *manager) Set(container *configs.Config) error {
...
 	// pids (since kernel 4.5)
-	if _, ok := m.controllers["pids"]; ok {
```
Comment (kolyshkin) [inline libcontainer/cgroups/fs2/fs2.go:180]: I believe this function was the only user of `controllers` map. If there aren't any left, remove it altogether.

Maybe as a separate patch, describing why there's no need for this map and check.

File: libcontainer/cgroups/fs2/memory.go (hunk 2 above)
Comment (kolyshkin) [reply libcontainer/cgroups/fs2/memory.go:53]: Can we have this fix as a separate patch? And I still don't understand why we haven't caught it before.

**Comment by kolyshkin** [comment]
Left a number of comments. It's better to split this into a few smaller patches as there are a few independent things being fixed.

**Comment by lifubang** [comment]
> It's better to split this into a few smaller patches as there are a few independent things being fixed.

I quite agree with you. And the purpose of this patch is to let `runc update` work.
After this fix, you can do more refactoring work.

File: libcontainer/cgroups/fs2/memory.go (hunk 2 above)
Comment (lifubang) [reply libcontainer/cgroups/fs2/memory.go:53]: We don’t check memorySwap==0

Comment (lifubang) [reply libcontainer/cgroups/fs2/memory.go:53]: @kolyshkin Do we still need to submit a separate patch?

File: libcontainer/cgroups/fs2/fs2.go (hunk 7 above)
Comment (lifubang) [reply libcontainer/cgroups/fs2/fs2.go:180]: agree with you

File: libcontainer/cgroups/fs2/memory.go (hunk 2 above)
Comment (lifubang) [reply libcontainer/cgroups/fs2/memory.go:53]: The seprate patch: #2405

File: libcontainer/cgroups/fs2/fs2.go (hunk 7 above)
Comment (lifubang) [reply libcontainer/cgroups/fs2/fs2.go:180]: Oh, there is another user: `GetStats`!

File: libcontainer/cgroups/fs2/memory.go (hunk 2 above)
Comment (lifubang) [reply libcontainer/cgroups/fs2/memory.go]: This part is fixed by #2410

**Comment by lifubang** [comment]
As #2390 and #2410 has been merged, so set this PR ready for review.

File: libcontainer/cgroups/fs2/create.go [hunk 8]
```diff
@@ -134,10 +138,11 @@ func CreateCgroupPath(path string, c *configs.Cgroup) (Err error) {
 		// enable needed controllers
```
Comment (AkihiroSuda) [inline libcontainer/cgroups/fs2/create.go:138]: this comment is no longer correct?

**Comment by AkihiroSuda** [comment]
LGTM, can we also have tests?

**Comment by lifubang** [comment]
> LGTM, can we also have tests?

Yes, I think we should add some test cases. I'll add it later.

File: tests/integration/cgroups.bats [hunk 9]
```diff
@@ -71,6 +72,24 @@ EOF
...
+    # run a detached busybox to work with
+    runc run -d --console-socket $CONSOLE_SOCKET test_cgroups_mem
+    [ "$status" -eq 0 ]
+
+    # update memory limit
+    runc update test_cgroups_mem --memory 50331648
+
+    # check value of memory.max
+    [ "$status" -eq 0 ]
+    check_cgroup_value "memory.max" 50331648
```
Comment (AkihiroSuda) [inline tests/integration/cgroups.bats:90]: Just compare `/sys/fs/cgroup/cgroup.controllers`.
On rootless systemd, the content should correspond to `/sys/fs/cgroup/user.slice/user-${uid}.slice/cgroup.controllers`

File: tests/integration/cgroups.bats [hunk 10]
```diff
@@ -110,6 +110,9 @@ EOF
 
     runc run -d --console-socket $CONSOLE_SOCKET test_cgroups_permissions
     [ "$status" -eq 0 ]
+    if [ "$CGROUP_UNIFIED" != "no" ]; then
+        check_cgroup_value "cgroup.controllers" "$(cat /sys/fs/cgroup/cgroup.controllers)"
```
Comment (AkihiroSuda) [inline tests/integration/cgroups.bats:114]: should be compared against `/sys/fs/cgroup/user.slice/user-${uid}.slice/cgroup.controllers` when rootless && `RUNC_USE_SYSTEMD` is set

File: libcontainer/cgroups/fs2/create.go (hunk 8 above)
Comment (AkihiroSuda) [reply libcontainer/cgroups/fs2/create.go:137]: ping @lifubang

Comment (lifubang) [reply libcontainer/cgroups/fs2/create.go]: changed now.

**Comment by AkihiroSuda** [comment]
LGTM (if green)

File: libcontainer/cgroups/fs2/create.go [hunk 11]
```diff
@@ -10,57 +11,58 @@ import (
...
 	}
 
-	// add the controller if available
-	add := func(controller string) {
+	// check whether the controller if available or not
+	have := func(controller string) bool {
 		if _, ok := avail[controller]; ok {
-			list = append(list, "+"+controller)
+			return true
 		}
```
Comment (kolyshkin) [inline libcontainer/cgroups/fs2/create.go:42]: nit
```go
_, ok := avail[controller]
return ok
```

File: libcontainer/cgroups/fs2/fs2.go [hunk 12]
```diff
@@ -76,8 +76,8 @@ func (m *manager) Apply(pid int) error {
 		// - "runc create (rootless + limits + no cgrouppath + no permission) fails with informative error"
 		if m.rootless {
 			if m.config.Path == "" {
-				cl, clErr := neededControllers(m.config)
-				if clErr == nil && len(cl) == 0 {
+				cl, clErr := needAnyControllers(m.config)
```
Comment (kolyshkin) [inline libcontainer/cgroups/fs2/fs2.go:79]: nit: `cl` used to mean "controller list". Now it's not so maybe rename to `need` or something.

ditto for `clErr`.

File: tests/integration/cgroups.bats [hunk 13]
```diff
@@ -110,6 +110,17 @@ EOF
 
     runc run -d --console-socket $CONSOLE_SOCKET test_cgroups_permissions
     [ "$status" -eq 0 ]
+    if [ "$CGROUP_UNIFIED" != "no" ]; then
+        if [ -n "${RUNC_USE_SYSTEMD}" ] ; then
```
Comment (kolyshkin) [inline tests/integration/cgroups.bats:114]: nit:
```go
if [ "$CGROUP_UNIFIED" != "no" -a -n "${RUNC_USE_SYSTEMD}" ]; then
```

Comment (lifubang) [reply tests/integration/cgroups.bats:114]: I think no, because there is `else` for `if [ -n "${RUNC_USE_SYSTEMD}" ] ; then`.

Comment (kolyshkin) [reply tests/integration/cgroups.bats:114]: sorry, missed it

**Comment by kolyshkin** [comment]
LGTM

@mrunalp @AkihiroSuda PTAL

**Comment by AkihiroSuda** [comment]
LGTM

**Pull request merged**